packages/firestore/src/util/bundle_reader.ts

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  BundleElement,
+  BundleMetadata
+} from '../protos/firestore_bundle_proto';
+import { Deferred } from './promise';
+import { AsyncQueue } from './async_queue';
+
+/**
+ * A complete element in the bundle stream, together with the byte length it
+ * occupies in the stream.
+ */
+export class SizedBundleElement {
+  constructor(
+    public readonly payload: BundleElement,
+    // How many bytes this element takes to store in the bundle.
+    public readonly byteLength: number
+  ) {}
+
+  isBundleMetadata(): boolean {
+    return 'metadata' in this.payload;
+  }
+}
+
+/**
+ * Create a `ReadableStream` from a underlying buffer.
+ *
+ * @param data: Underlying buffer.
+ * @param bytesPerRead: How many bytes to read from the underlying buffer from
+ * each read through the stream.
+ */
+export function toReadableStream(
+  data: Uint8Array | ArrayBuffer,
+  bytesPerRead = 10240
+): ReadableStream<Uint8Array | ArrayBuffer> {
+  let readFrom = 0;
+  return new ReadableStream({
+    start(controller) {},
+    async pull(controller): Promise<void> {
+      controller.enqueue(data.slice(readFrom, readFrom + bytesPerRead));
+      readFrom += bytesPerRead;
+      if (readFrom >= data.byteLength) {
+        controller.close();
+      }
+    }
+  });
+}
+
+/**
+ * A class representing a bundle.
+ *
+ * Takes a bundle stream or buffer, and presents abstractions to read bundled
+ * elements out of the underlying content.
+ */
+export class BundleReader {
+  /** Cached bundle metadata. */
+  private metadata: Deferred<BundleMetadata> = new Deferred<BundleMetadata>();
+  /** The reader instance of the given ReadableStream. */
+  private reader: ReadableStreamDefaultReader;
+  /**
+   * Internal buffer to hold bundle content, accumulating incomplete element
+   * content.
+   */
+  private buffer: Uint8Array = new Uint8Array();
+  /** The decoder used to parse binary data into strings. */
+  private textDecoder = new TextDecoder('utf-8');
+
+  constructor(
+    private bundleStream:
+      | ReadableStream<Uint8Array | ArrayBuffer>
+      | Uint8Array
+      | ArrayBuffer,
+    /** Async queue used to perform bundle reading. */
+    private asyncQueue: AsyncQueue = new AsyncQueue()
+  ) {
+    if (
+      bundleStream instanceof Uint8Array ||
+      bundleStream instanceof ArrayBuffer
+    ) {
+      this.bundleStream = toReadableStream(bundleStream);
+    }
+    this.reader = (this.bundleStream as ReadableStream).getReader();
+
+    this.nextElement().then(
```

Comment:
Maybe add a small inline comment before this line:

// Read the metadata (which is the first element)